package RecastIO;

message Vec3 
{
	optional float x							= 1;
	optional float y							= 2;
	optional float z							= 3;
}

message AxisAlignedBounds
{
	required Vec3		mins					= 1;
	required Vec3		maxs					= 2;
}

message NavMeshParams
{
	required float			agentHeightStand	= 1;
	required float			agentHeightCrouch	= 2;
	required float			agentRadius			= 3;
	required float			agentClimb			= 4;
	required float			walkSlopeAngle		= 5;
	required float			cellSize			= 6;
	required float			cellHeight			= 7;
	required float			edgeMaxLength		= 8;
	required float			edgeMaxError		= 9;
	required int32			tileSize			= 10;
	required float			detailSampleDist	= 11;
	required float			detailSampleMaxErr	= 12;
}

message Tile
{
	required int32			unCompressedSize	= 1;
	required bytes			compressedData		= 2;
}

message OffMeshConnection
{
	required Vec3			entryPos			= 1;
	required Vec3			exitPos				= 2;
	repeated Vec3			vertices			= 3;
	required float			radius				= 4;
	required uint64			areaflags			= 5;
	optional bool			biDirectional		= 6 [default=false];
}

message Material
{
	required string			name				= 1;
	optional uint32			contents			= 2;
	optional uint32			surfaceflags		= 3;
}

message Model
{
	message TriParms
	{
		required uint32		tri					= 1;
		optional uint32		surfaceOverride		= 2;
	}

	required string			name				= 1;
	required uint32			modelcrc			= 2;
	required uint32			numMeshTris			= 3;
	repeated TriParms		triangles			= 4;
}

enum ShapeMode
{
	SHAPE_TRIANGLES		= 0;
	SHAPE_OBB			= 1;
	//SHAPE_CONVEXHULL	= 2;
};

message NodeState
{
	oneof type {
		int32			subModelId				= 1;
		int32			staticModelId			= 2;
		int32			displacementId			= 3;
	}
	
	optional bool		enabled					= 5 [default=true];
	optional ShapeMode	shapemode				= 6 [default=SHAPE_TRIANGLES];
	optional bool		solid					= 7 [default=true];
	optional bool		dynamic					= 8 [default=false];
	optional uint64		navFlagOverride			= 9 [default=0];
	optional uint32		activeModelCrc			= 10 [default=0];
	optional string		modelname				= 11;
	optional string		name					= 12;
}

message NavigationMesh 
{
	required int32				version				= 1;
	required NavMeshParams		navMeshParams		= 2;	
	repeated AxisAlignedBounds	exclusionZone		= 3;
	repeated OffMeshConnection	offMeshConnection	= 4;
	repeated NodeState			nodeState			= 5;
	repeated Model				models				= 6;

	repeated Tile				tiles				= 100;
}
